Update 00030-Functionalities.md (#18)
oskariorg/oskari-documentation: 4 Frontend main functionalities/00030-Functionalities.md

## Map layer selector

The map layers can be accessed via map layer selector. The Map layer selector contains a list of all available map layers in the Oskari instance, grouped either by theme or by data provider. The map layer selector can be used to show map layers, filter them and to access their metadata. To view metadata there is link to the metadata display module, which accesses CSW interface.

Supported layer types are:
- map layer
- map layer stack
- background map layer
- WFS layer
- thematic map layer
- analysis results layer
- user's own data layer
- time-enabled WMS layer (WMS-T)

If needed, new map layer types can be also added programmatically.

## Selected map layers window

Oskari UI has a separate window for the selected map layers. Here are listed all the layers displayed in the map view.

The functionalities of Selected map layers window are:

- Layers have localized titles (and optionally subtitles)
- Layers can be temporarily hidden and shown
- Layers can be organized in the layer stack by dragging and dropping
- Layer opacity can be controlled using a slidebar or by entering opacity percentage

> - Layer metadata can be viewed (link to the metadata display portlet, which accesses CSW interface)

Reviewer: Portlet refers to time where PTI was implemented on Liferay portal product. Maybe something like this?
```suggestion
- Layer metadata can be viewed (fetched for display using CSW interface)
```

- WFS layers provide a link to attribute information table
- WFS layer style can be changed on a temporary basis
- GeoServer backed WMS layers can be visualized as heatmaps
- For WMS layers the available styles can be viewed and selected for use

## Menu and toolbar

> The whole Oskari UI can be personalized by the main users. This includes the menu on the side of the map window. The menu can be, for example, hidden and the individual menu entries can be configured individually.

Reviewer: Should we find a better term for "main users"? Administrators? But do we want to separate technical config from "admins" that mostly work through the UI?

Author: That's what I thought, too. Can also be "The whole Oskari UI can be customized." without mentioning the main user / admin / developer 🤔

Reviewer: That sounds way better with removal of main users and personalized -> customized.

Toolbar is anchored to the menu bar and contains various tools. The toolbar can be hidden and different buttons can be disabled and enabled individually. Tools can also be made to appear when a specific map layer is added to selected layers list.

> **The toolbar buttons include**:

Reviewer: Technically the toolbar is something that _any_ functionality can add buttons to. So one could have an application-specific functionality that has a button on the toolbar. Maybe something like out of the box these are the functionalities that are available and applications can add to them?

Author: Yes! 👍

- Drag & pan button (enabled by default)
- Rubber band zoom button
- Restore initial map state button
- Map view history manager: navigation with back and forward buttons
- Measurement tools: measure distance and area
- Link map view: creates an URL which opens the current map view
- Marker tool: markers with associated text can be created on the map and sent as link
- Print map view: starts the print mode
- Save map view: Map view configurations can be saved by logged-in user

> - Find nearest place tool: Find the nearest place name by clicking on the map

Reviewer: Requires a compatible search backend that supports reverse geocoding. One is included in the sample app though, but for anyone considering this it would be good to know that you can actually decide what is used as a backend for this?

## Location search and metadata search

Oskari has search functions for certain locations and the metadata of map layers.

> **The location search** is a one-field search which can return search results from multiple sources, such as placename, address, cadastral parcel or similar services. The service access parameters are configurable in the backend.

Reviewer: Again, just using the demo you can't really search for cadastral parcels on it, BUT you can configure a search backend that does support this (or any other type of search result that would be a better fit for some application).

- Layers can be temporarily hidden and shown
- Layers can be organized in the layer stack by dragging and dropping
- Layer opacity can be controlled using a slidebar or by entering opacity percentage
- Layer metadata can be viewed (fetched for display using CSW interface)

> - WFS layers provide a link to attribute information table

Reviewer 2: Improve terminology by replacing "attribute information table" by "feature data" (as n the UI) or by "feature property data" (as in the admin-panel). 
![image](https://github.com/user-attachments/assets/f434815d-91a0-46df-8e78-a927cea32330)
![image](https://github.com/user-attachments/assets/0c0a8610-d435-4aec-bc06-fbcfb33bce8c)
